apps/src/weblab/brambleHost.js

```diff
@@ -302,6 +302,66 @@ function syncFilesWithBramble(fileEntries, currentProjectVersion, callback) {
   }
 }
 
+function validateProjectChanged(callback) {
+  getAllFileDataFromBramble((userFiles, error) => {
+    const startSources = webLab_.getStartSources();
+
+    // Don't let an error from bramble block the student from progressing.
+    if (error || userFiles.files.length !== startSources.files.length) {
+      callback(true /* project changed */);
+      return;
+    }
+
+    const changedFile = startSources.files.find(startFile => {
+      const matchingFile = userFiles.files.find(
+        file => file.name === startFile.name
+      );
+      // If startFile doesn't have `data` (and instead has something different
+      // like `url`), this is an image and is stored differently in the
+      // startSources than in bramble. Check for a matching file name, but
+      // don't compare data.
```

Comment:
Does this refer to startFile containing the string `data` vs. `url`, or is it more like binary content vs. alphanumeric content?

Reply:
Both. If startFile contains 'url' then that means the source in bramble will be stored as 'data' with binary content. If startFile contains 'data' then that means the source in bramble will be stored as 'data' with alphanumeric content.